Data Preparation

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [ ]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

In [ ]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [ ]:
n_cols = predictors_norm.shape[1] # number of predictors

A. Build a baseline model
Use the Keras library to build a neural network with the following:
- One hidden layer of 10 nodes, and a ReLU activation function
- Use the adam optimizer and the mean squared error  as the loss function.

In [ ]:
import keras

In [ ]:
from keras.models import Sequential
from keras.layers import Dense

In [ ]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))

    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [ ]:
# build the model
model = regression_model()

1. Randomly split the data into a training and test sets by holding 30% of the data for testing.

In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.30, random_state=42)

2. Train the model on the training data using 50 epochs.

In [ ]:
# fit the model
model.fit(X_train, y_train, epochs=100, verbose=2)

Epoch 1/100
23/23 - 2s - loss: 1633.3750 - 2s/epoch - 96ms/step
Epoch 2/100
23/23 - 0s - loss: 1618.4097 - 101ms/epoch - 4ms/step
Epoch 3/100
23/23 - 0s - loss: 1603.9104 - 82ms/epoch - 4ms/step
Epoch 4/100
23/23 - 0s - loss: 1589.6324 - 86ms/epoch - 4ms/step
Epoch 5/100
23/23 - 0s - loss: 1575.1234 - 132ms/epoch - 6ms/step
Epoch 6/100
23/23 - 0s - loss: 1560.4014 - 180ms/epoch - 8ms/step
Epoch 7/100
23/23 - 0s - loss: 1545.3599 - 122ms/epoch - 5ms/step
Epoch 8/100
23/23 - 0s - loss: 1529.8015 - 152ms/epoch - 7ms/step
Epoch 9/100
23/23 - 0s - loss: 1513.5496 - 123ms/epoch - 5ms/step
Epoch 10/100
23/23 - 0s - loss: 1496.4279 - 139ms/epoch - 6ms/step
Epoch 11/100
23/23 - 0s - loss: 1478.3805 - 196ms/epoch - 9ms/step
Epoch 12/100
23/23 - 0s - loss: 1459.0198 - 142ms/epoch - 6ms/step
Epoch 13/100
23/23 - 0s - loss: 1438.7909 - 160ms/epoch - 7ms/step
Epoch 14/100
23/23 - 0s - loss: 1417.2960 - 131ms/epoch - 6ms/step
Epoch 15/100
23/23 - 0s - loss: 1394.6031 - 157ms/epoch - 7ms/step
Epoch 16

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength.

In [ ]:
y_pred = model.predict(X_test)

10/10 [==============================] - 0s 2ms/step


In [ ]:
from sklearn.metrics import mean_squared_error

In [ ]:
mse = mean_squared_error(y_test, y_pred)

In [ ]:
mse

170.34035101181087

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

In [ ]:
mse_list = []
for i in range(50):
  X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.30, random_state=42)
  model.fit(X_train, y_train, epochs=100)
  y_pred = model.predict(X_test)
  mse = mean_squared_error(y_test, y_pred)
  mse_list.append(mse)

Strumieniowane dane wyjściowe obcięte do 5000 ostatnich wierszy.
Epoch 14/100
23/23 [==============================] - 0s 3ms/step - loss: 112.1925
Epoch 15/100
23/23 [==============================] - 0s 3ms/step - loss: 114.9874
Epoch 16/100
23/23 [==============================] - 0s 3ms/step - loss: 112.8297
Epoch 17/100
23/23 [==============================] - 0s 3ms/step - loss: 120.2759
Epoch 18/100
23/23 [==============================] - 0s 3ms/step - loss: 117.8015
Epoch 19/100
23/23 [==============================] - 0s 3ms/step - loss: 113.1115
Epoch 20/100
23/23 [==============================] - 0s 3ms/step - loss: 116.2488
Epoch 21/100
23/23 [==============================] - 0s 3ms/step - loss: 126.8926
Epoch 22/100
23/23 [==============================] - 0s 3ms/step - loss: 111.5192
Epoch 23/100
23/23 [==============================] - 0s 3ms/step - loss: 130.1084
Epoch 24/100
23/23 [==============================] - 0s 3ms/step - loss: 129.6675
Epoch 25/100
23/23 [==

5. Report the mean and the standard deviation of the mean squared errors.

In [ ]:
mse_list

[4649.218593962709,
 1177.2842449570578,
 268.10313945199994,
 112.3579131334886,
 118.92305964621193,
 113.79878354982183,
 113.63071345661226,
 125.23610509031668,
 110.28281727676523,
 111.13357741227092,
 110.54343391798956,
 139.3789136386974,
 111.87414263012646,
 111.57768023382275,
 137.41701832005774,
 114.03917444144864,
 142.19643733243848,
 120.6717191513407,
 117.26050672267549,
 109.14747860035334,
 110.1015801706184,
 115.89526320670817,
 122.09772847890538,
 157.0405675832852,
 113.18604010031792,
 120.30950741389645,
 120.41606815288718,
 110.58153079304795,
 109.18682345632271,
 108.47073521347289,
 116.20365583552913,
 108.69129689823237,
 135.97147421004138,
 108.14159733611396,
 108.42465589055368,
 110.52986557081012,
 109.3932840600294,
 119.88810243817233,
 107.39250054723834,
 157.64602777221177,
 108.61541695561529,
 121.84658423142766,
 161.67925632850938,
 133.1399770993147,
 120.51622760615115,
 127.64295632217531,
 134.51229034483208,
 109.62270609639414,


In [ ]:
len(mse_list)

50

In [ ]:
mean = sum(mse_list)/len(mse_list)

In [ ]:
mean

234.40199446103287

In [ ]:
std_dev_list = abs(mse_list - mean)
std_dev = sum(std_dev_list)/len(std_dev_list)

In [ ]:
std_dev

215.65599979954663

How does the mean of the mean squared errors compare to that from Step B?

Part B:
mean mse from 50 models: 497.17

Part C: mean mse from 50 models: 234.40

The result in part C is better